hw/xbox/mcpx_apu.c

```diff
@@ -461,7 +465,34 @@ static uint64_t gp_read(void *opaque,
 {
     MCPXAPUState *d = opaque;
 
-    uint64_t r = d->gp.regs[addr];
+    assert(size == 4);
+
+    uint64_t r = 0;
+    switch (addr) {
+    case NV_PAPU_GPXMEM ... NV_PAPU_GPXMEM + 0xFFF*4: {
+        uint32_t xaddr = (addr - NV_PAPU_GPXMEM) / 4;
+        r = dsp_read_memory(d->gp.dsp, 'X', xaddr);
+        break;
+    }
+    case NV_PAPU_GPMIXBUF ... NV_PAPU_GPMIXBUF + 0x3FF*4: {
+        uint32_t xaddr = (addr - NV_PAPU_GPMIXBUF) / 4;
+        r = dsp_read_memory(d->gp.dsp, 'X', 0x001400 + xaddr);
+        break;
+    }
+    case NV_PAPU_GPYMEM ... NV_PAPU_GPYMEM + 0x7FF*4: {
+        uint32_t yaddr = (addr - NV_PAPU_GPYMEM) / 4;
+        r = dsp_read_memory(d->gp.dsp, 'Y', yaddr);
+        break;
+    }
+    case NV_PAPU_GPPMEM ... NV_PAPU_GPPMEM + 0xFFF*4: {
+        uint32_t paddr = (addr - NV_PAPU_GPPMEM) / 4;
+        r = dsp_read_memory(d->gp.dsp, 'P', paddr);
+        break;
+    }
+    default:
+        r = d->gp.regs[addr];
```

Comment:
Prefer to `assert` here that addr is not out of bounds

Reply:
This was copied from the existing `gp_write` / `ep_read`. Neither of those do any checks.

Let's do it in a cleanup later. I don't think this will happen anyway.